src/main/scala/org/constellation/p2p/Cluster.scala

```diff
@@ -463,18 +461,34 @@ class Cluster[F[_]: Concurrent: Logger: Timer: ContextShift](ipManager: IPManage
       "cluster_leave"
     )
 
-  def broadcastNodeState(): F[Unit] =
+  private def broadcastNodeState(nodeState: NodeState): F[Unit] =
     logThread(
-      getNodeState.flatMap { nodeState =>
-        broadcast(_.postNonBlockingUnitF("status", SetNodeStatus(dao.id, nodeState))(C))
-      }.flatTap {
+      broadcast(_.postNonBlockingUnitF("status", SetNodeStatus(dao.id, nodeState))(C)).flatTap {
         _.filter(_._2.isLeft).toList.traverse {
           case (id, e) => Logger[F].warn(s"Unable to propagate status to node ID: $id")
         }
       }.void,
       "cluster_broadcastNodeState"
     )
 
+//  def compareAndSet(expected: NodeState, newState: NodeState, skipBroadcast: Boolean = false): F[SetStateResult] =
+//    compareAndSet(Set(expected), newState, skipBroadcast)
+
+  def compareAndSet(expected: Set[NodeState], newState: NodeState, skipBroadcast: Boolean = false): F[SetStateResult] =
+    nodeState.modify { current =>
+      if (expected.contains(current)) (newState, SetStateResult(current, isNewSet = true))
+      else (current, SetStateResult(current, isNewSet = false))
+    }.flatTap(
+        res =>
+          if (res.isNewSet) LiftIO[F].liftIO(dao.metrics.updateMetricAsync[IO]("nodeState", newState.toString))
+          else Sync[F].unit
+      )
+      .flatTap {
+        case SetStateResult(_, false)                                      => Sync[F].unit
+        case SetStateResult(_, true) if skipBroadcast                      => Sync[F].unit
+        case SetStateResult(_, true) if broadcastStates.contains(newState) => broadcastNodeState(newState)
```

Comment:
why are you using `skipBroadcast` flag if anyway you're checking for broadcastStates?